src/message/pack_encrypted/authcrypt.rs

```diff
@@ -0,0 +1,326 @@
+                    AnonCryptAlg::A256cbcHs512EcdhEsA256kw => jwe::encrypt::<
+                        AesKey<A256CbcHs512>,
+                        EcdhEs<'_, X25519KeyPair>,
+                        X25519KeyPair,
+                        AesKey<A256Kw>,
+                    >(
+                        msg.as_bytes(),
+                        jwe::Algorithm::EcdhEsA256kw,
+                        jwe::EncAlgorithm::A256cbcHs512,
+                        None,
+                        &to_keys,
+                    )
+                    .context("Unable produce authcrypt envelope")?,
+                    AnonCryptAlg::Xc20pEcdhEsA256kw => jwe::encrypt::<
+                        Chacha20Key<XC20P>,
+                        EcdhEs<'_, X25519KeyPair>,
+                        X25519KeyPair,
+                        AesKey<A256Kw>,
+                    >(
+                        msg.as_bytes(),
+                        jwe::Algorithm::EcdhEsA256kw,
+                        jwe::EncAlgorithm::Xc20P,
+                        None,
+                        &to_keys,
+                    )
+                    .context("Unable produce authcrypt envelope")?,
+                    AnonCryptAlg::A256gcmEcdhEsA256kw => jwe::encrypt::<
+                        AesKey<A256Gcm>,
+                        EcdhEs<'_, X25519KeyPair>,
+                        X25519KeyPair,
+                        AesKey<A256Kw>,
+                    >(
+                        msg.as_bytes(),
+                        jwe::Algorithm::EcdhEsA256kw,
+                        jwe::EncAlgorithm::A256Gcm,
+                        None,
+                        &to_keys,
+                    )
+                    .context("Unable produce authcrypt envelope")?,
+                }
+            } else {
+                msg
+            }
+        }
+        KnownKeyAlg::P256 => {
+            let _to_keys = to_keys
+                .iter()
+                .map(|vm| vm.as_p256().map(|k| (&vm.id, k)))
+                .collect::<Result<Vec<_>>>()?;
+
+            let to_keys: Vec<_> = _to_keys
+                .iter()
+                .map(|(id, key)| (id.as_str(), key))
+                .collect();
+
+            let msg = match enc_alg_auth {
+                AuthCryptAlg::A256cbcHs512Ecdh1puA256kw => jwe::encrypt::<
+                    AesKey<A256CbcHs512>,
+                    Ecdh1PU<'_, P256KeyPair>,
+                    P256KeyPair,
+                    AesKey<A256Kw>,
+                >(
+                    msg,
+                    jwe::Algorithm::Ecdh1puA256kw,
+                    jwe::EncAlgorithm::A256cbcHs512,
+                    Some((&from_key.id, &from_priv_key.as_p256()?)),
+                    &to_keys,
+                )
+                .context("Unable produce authcrypt envelope")?,
+            };
+
+            if protect_sender {
+                match enc_alg_anon {
+                    AnonCryptAlg::A256cbcHs512EcdhEsA256kw => jwe::encrypt::<
```

Comment:
Looks like a candidate for a helper function (something like `anoncrypt(to_keys)`)